File: builder/dart/builder.dart
main() {
  // using builder with director class
  final carBuilder = CarBuilder();
  final director = Director(carBuilder);
  final deportiveBuilder = (director.getDeportiveCarBuilder() as CarBuilder).build();
  final truckBuilder = (director.getTruckBuilder() as CarBuilder).build();

  print(deportiveBuilder.toString());

  print(truckBuilder.toString());

  // using only builder

  final manualBuilder = ManualBuilder();

  final manual = (manualBuilder.setEngine("Deportive").makeDoors(2).setWheels(4)
          as ManualBuilder)
      .build();
  print(manual);
}

class Automovil {
  Automovil({this.engine = "Standar", this.doors = 4, this.wheel = 4});
  String engine;
  int doors;
  int wheel;

  @override
  String toString() {
    return "Automovil( engine = $engine, doors = $doors, whell = $wheel )";
  }
}

class Manual {
  final List<String> descriptions;

  Manual({required List<String> this.descriptions});

  @override
  String toString() {
    return "Manual( descriptions = $descriptions)";
  }
}

abstract class Builder {
  reset();
  Builder makeDoors(int doors);
  Builder setWheels(int wheel);
  Builder setEngine(String engine);
}

class CarBuilder extends Builder {
  Automovil car = Automovil();
  @override
  reset() {
    car = Automovil();
  }

  @override
  Builder makeDoors(int doors) {
    car.doors = doors;
    return this;
  }

  @override
  Builder setWheels(int wheel) {
    car.wheel = wheel;
    return this;
  }

  @override
  Builder setEngine(String engine) {
    car.engine = engine;
    return this;
  }

  Automovil build() {
    // Reset objet is optional when is build
    final automovil = car;
    reset();
    return automovil;
  }
}

class ManualBuilder extends Builder {
  Manual manual = Manual(descriptions: []);
  @override
  reset() {
    manual = Manual(descriptions: []);
  }

  @override
  Builder makeDoors(int doors) {
    manual.descriptions.add("Car has $doors doors");
    return this;
  }

  @override
  Builder setWheels(int wheel) {
    manual.descriptions.add("Car has $wheel wheels");
    return this;
  }

  @override
  Builder setEngine(String engine) {
    manual.descriptions.add("Car has $engine engines");
    return this;
  }

  Manual build() {
    return manual;
  }
}

class Director {
  final Builder builder;
  Director(this.builder);
  Builder getDeportiveCarBuilder() {
    return builder.setEngine("Deportive").makeDoors(2).setWheels(4);
  }

  Builder getTruckBuilder() {
    return builder.setEngine("big").makeDoors(2).setWheels(8);
  }
}

File: builder/python/builder.py
from abc import ABC, abstractmethod


    
class Automovil:
    
    def __init__(self,engine: str = "Standar", doors: int = 4, wheel: int = 4, trailer = False):
        self.engine = engine
        self.doors = doors
        self.wheel = wheel
        self.trailer = trailer
        
    def __str__(self) -> str:
        return f"Automovil( engine = {self.engine}, doors = {self.doors}, wheel = {self.wheel} trailer = {self.trailer})"

class Manual:
    def __init__(self, descriptions: list[str] = []):
        self.descriptions = descriptions
    
    def __str__(self) -> str:
        return f"Manual ( descriptions = {self.descriptions} )"

class Builder(ABC):
    @abstractmethod
    def reset(self):
        pass
    
    @abstractmethod
    def makeDoors(self,doors: int):
        pass
    
    @abstractmethod
    def setWheels(self,wheel: int):
        pass
    
    @abstractmethod
    def setEngine(self,engine: str):
        pass
    
    @abstractmethod
    def addTrailer(self):
        pass
    


class CarBuilder(Builder):
    car: Automovil = Automovil()
    
    def reset(self):
        self.car = Automovil()
    

    def makeDoors(self,doors: int)-> Builder:
        self.car.doors = doors
        return self
    

    def setWheels(self,wheel: int)-> Builder:
        self.car.wheel = wheel
        return self
    

    def setEngine(self,engine: str)-> Builder:
        self.car.engine = engine
        return self
    
    def addTrailer(self):
        self.car.trailer = True
        return self

    def build(self)-> Automovil:
        # Reset objet is optional when is build
        automovil = self.car
        self.reset()
        return automovil
    
class ManualBuilder(Builder):
    manual: Manual = Manual()
    
    def reset(self):
        self.manual.descriptions = []
    

    def makeDoors(self,doors: int)-> Builder:
        self.manual.descriptions.append(f"Car has {doors} doors")
        return self
    

    def setWheels(self,wheel: int)-> Builder:
        self.manual.descriptions.append(f"Car has {wheel} wheel")
        return self
    

    def setEngine(self,engine: str)-> Builder:
        self.manual.descriptions.append(f"Car has {engine} engine")
        return self
    
    def addTrailer(self):
        self.manual.descriptions.append("Car has trailer")
        return self
    

    def build(self)-> Automovil:
        return self.manual
    




class Director:
    def __init__(self,builder: Builder):
        self.builder  = builder

    def getDeportiveCarBuilder(self)-> Builder:
        return self.builder.setEngine("Deportive").makeDoors(2).setWheels(4)
    

    def getTruckBuilder(self)->Builder:
        return self.builder.setEngine("big").makeDoors(2).setWheels(8).addTrailer()


    
def main():
    # using builder with director class
    carBuilder = CarBuilder()
    director = Director(carBuilder)
    deportiveBuilder = director.getDeportiveCarBuilder().build()
    truckBuilder = director.getTruckBuilder().build()

    print(deportiveBuilder)

    print(truckBuilder)

    #using only builder

    manualBuilder = ManualBuilder()

    manual = manualBuilder.setEngine("Deportive").makeDoors(2).setWheels(4)
    print(manual.build())
    
    manualBuilder.reset()
    
    
    manualTruck = manualBuilder.setEngine("big").makeDoors(4).setWheels(8).addTrailer()
    print(manualTruck.build())


if __name__ == '__main__': 
    main()